### PR TITLE
Plugin: Dedicated Page for Rarely Used Bookmarks

### PR DESCRIPTION
**Pull request by leet1994**
First of all, I am not sure if this is there right location where you want plugins to live in. I have not found any HowTo article for plugin development nor the integration into the plugin marketplace, so I decided to put it directly into the `api/plugins` folder.

### Problem Description
I personally use Organizr to collect links to all of my self-hosted services as well as some useful sites. Some of the latter ones are used on a daily basis and some are niche-like things that I only need once in a few weeks. Although they are used so rarely, I do not want to forget about them and maybe use them on all of my devices.
I thought there was a lack of utility for this kind of links that do not deserve to be placed in the sidebar based on their seldom use.

### Solution
For this scenario I created a plugin called simply `Bookmark` that creates a new API endpoint to retrieve a configurable bookmark collection site. The site can be added with type `Organizr` to the sidebar as usual. It will then create a nice looking site like this:
grafik

When activating the plugin, it will add two new sub pages to the `Tab Editor` menu. These are very similar to the `Tabs` and `Categories` sub pages but dedicated for the `Bookmark` plugin. The categories and tabs to show on the bookmark page can be configured as usual on these pages.

A detailed description what you have to do is available in the settings menu of the plugin and covers:
* Add tab that points to `api/v2/plugins/bookmark/page` and set it's type to `Organizr`
* Create Bookmark categories in the new area in `Tab Editor`
* Create Bookmark tabs in the new area in `Tab Editor`
* Open your custom Bookmark page via menu

Although there is potential for improvements, I hope you like the plugin :)

### CODE REVIEW

**Comment by causefx** [comment]
I'll go thru the code here shortly and we can discuss. cool idea :)

**Comment by causefx** [comment]
I'm reworking the binding of plugins js stuff - we can go through the changes here shortly

**Comment by causefx** [comment]
@leet1994 can you rename `api/plugins/js/bookmark.js` to `api/plugins/js/bookmark-settings.js` to reflect the new changes I pushed today?

File: .gitignore [hunk 1]
```diff
@@ -156,6 +156,10 @@ api/plugins/*
 !api/plugins/misc/speedTest/telemetry.php
 !api/plugins/misc/speedTest/telemetry_settings.php
 !api/plugins/misc/speedTest/speedtest_worker.min.js
+!api/plugins/bookmark.php
+!api/plugins/api/bookmark.php
+!api/plugins/config/bookmark.php
+!api/plugins/js/bookmark.js
```
Comment (causefx) [inline .gitignore]: update to:
`!api/plugins/js/bookmark-settings.js`

File: api/plugins/api/bookmark.php [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+	}
+	$response->getBody()->write(jsonE($GLOBALS['api']));
+	return $response
+		->withHeader('Content-Type', 'application/json;charset=UTF-8')
+		->withStatus($GLOBALS['responseCode']);
+});
+$app->get('/plugins/bookmark/tabs/{id}', function ($request, $response, $args) {
+	$Bookmark = new Bookmark();
+	if ($Bookmark->_checkRequest($request) && $Bookmark->checkRoute($request)) {
+		if ($Bookmark->qualifyRequest(1, true)) {
```
Comment (causefx) [inline api/plugins/api/bookmark.php]: You can remove this if statement as you are checking user in your function, leaving as is will only allow admins and co admins to access.

File: api/plugins/bookmark.php [hunk 3]
```diff
@@ -0,0 +1,1082 @@
...
+                    </div>
+				</div>
+			'
+		);
+	}
+
+	public function _getPage()
+	{
+		$script = '
+<script>
```
Comment (causefx) [inline api/plugins/bookmark.php]: if you like, you could move all this into an actual css file and it will get included if plugin is enabled.

**Comment by leet1994** [comment]
Hey, thanks for the review so far :)
I'll make the suggested changes this weekend.

**Comment by causefx** [comment]
sounds good!

**Comment by leet1994** [comment]
@causefx implemented the changes :)
I had to back-merge from v2-develop to get your recent changes, I hope that's no problem.

**Pull request merged**